compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -4036,22 +4038,25 @@ fn print_disambiguation_help<'tcx>(
             .into_iter()
             .chain(args)
             .map(|arg| {
-                tcx.sess.source_map().span_to_snippet(arg.span).unwrap_or_else(|_| "_".to_owned())
+                tcx.sess
+                    .source_map()
+                    .span_to_snippet(arg.span)
+                    .map(Cow::Owned)
+                    .unwrap_or("_".into())
             })
             .collect::<Vec<_>>()
             .join(", ");
 
             let args = format!("({}{})", rcvr_ref, args);
+            let candidate = if let Some(candidate) = candidate_idx {
+                &format!("candidate #{candidate}")
+            } else {
+                "the candidate"
+            };
+
             err.span_suggestion_verbose(
                 span,
-                format!(
-                    "disambiguate the {def_kind_descr} for {}",
-                    if let Some(candidate) = candidate_idx {
-                        format!("candidate #{candidate}")
-                    } else {
-                        "the candidate".to_string()
-                    },
-                ),
+                format!("disambiguate the {def_kind_descr} for {candidate}",),
```

Comment:
I think you could do the `write!` pattern you've used elsewhere to avoid the double `format!`. Roughly

```rust
let mut msg = format!("disambiguate the {def_kind_descr} for ");
match candidate_idx {
    Some(candidate) => write!(msg, "candidate #{candidate}"),
    None => write!(msg, "the candidate"),
}
```